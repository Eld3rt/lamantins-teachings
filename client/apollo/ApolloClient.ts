import { HttpLink } from '@apollo/client'
import { NextSSRInMemoryCache, NextSSRApolloClient } from '@apollo/experimental-nextjs-app-support/ssr'
import { registerApolloClient } from '@apollo/experimental-nextjs-app-support/rsc'
import { cookies } from 'next/headers'

export const { getClient } = registerApolloClient(() => {
  let authToken = cookies().get('sid')?.value

  return new NextSSRApolloClient({
    cache: new NextSSRInMemoryCache(),
    link: new HttpLink({
      uri: 'http://localhost:3000/graphql',
      headers: { Cookie: `${authToken ? `sid=${authToken}` : ''}` },
    }),
  })
})
