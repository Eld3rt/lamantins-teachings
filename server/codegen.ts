import type { CodegenConfig } from '@graphql-codegen/cli'

const config: CodegenConfig = {
  schema: './src/graphql/schema/*/index.ts',
  generates: {
    './src/graphql/types/resolvers-types.ts': {
      plugins: ['typescript', 'typescript-resolvers'],
      config: {
        useIndexSignature: true,
        emitLegacyCommonJSImports: false,
        contextType: '../../apollo/server#MyContext',
        mappers: {
          User: '.prisma/client#User as UserModel',
        },
      },
    },
  },
}
export default config
