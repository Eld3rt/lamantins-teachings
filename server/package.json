{
  "name": "server",
  "type": "module",
  "version": "1.0.0",
  "description": "",
  "main": "server.ts",
  "scripts": {
    "dev": "tsx watch index.ts",
    "codegen": "graphql-codegen-esm --config codegen.ts",
    "test": "jest --config jest.config.ts ./__tests__"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "@apollo/server": "^4.9.5",
    "@prisma/client": "^5.10.1",
    "bcrypt": "^5.1.1",
    "cookie-parser": "^1.4.6",
    "cors": "^2.8.5",
    "dotenv": "^16.3.1",
    "express": "^4.18.2",
    "graphql": "^16.8.1",
    "graphql-scalars": "^1.22.4",
    "graphql-tag": "^2.12.6",
    "ioredis": "^5.3.2",
    "jsonwebtoken": "^9.0.2",
    "lodash": "^4.17.21",
    "nodemailer": "^6.9.9",
    "uuid": "^9.0.1",
    "yup": "^1.3.3"
  },
  "devDependencies": {
    "@graphql-codegen/cli": "^5.0.0",
    "@graphql-codegen/typescript": "4.0.1",
    "@graphql-codegen/typescript-resolvers": "4.0.1",
    "@types/bcrypt": "^5.0.2",
    "@types/cookie-parser": "^1.4.7",
    "@types/cors": "^2.8.17",
    "@types/jest": "^29.5.11",
    "@types/jsonwebtoken": "^9.0.5",
    "@types/lodash": "^4.14.202",
    "@types/nodemailer": "^6.4.14",
    "@types/uuid": "^9.0.8",
    "jest": "^29.7.0",
    "prisma": "^5.10.1",
    "ts-jest": "^29.1.1",
    "tsx": "^4.7.0",
    "typescript": "^5.3.3"
  },
  "prisma": {
    "schema": "src/prisma/schema.prisma",
    "seed": "tsx src/prisma/seed.ts"
  }
}
