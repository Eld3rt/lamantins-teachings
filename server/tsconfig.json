{
  "compilerOptions": {
    "sourceMap": true,
    "outDir": "dist",
    "strict": true,
    "lib": ["esnext"],
    "esModuleInterop": true,
    "target": "es2020",
    "module": "Node16",
    "moduleResolution": "Node16"
  },
  "ts-node": {
    "esm": true
  }
}
