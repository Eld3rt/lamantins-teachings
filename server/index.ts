import express, { json } from 'express'
import cors from 'cors'
import { server } from './src/apollo/server'
import { prisma } from './src/prisma/prisma'
import { expressMiddleware } from '@apollo/server/express4'

const PORT = process.env.PORT || 5050
const app = express()

const corsOptions = {
  origin: process.env.BASE_URL,
  credentials: true,
}

app.use(cors(corsOptions))
app.use(express.json())

await server.start()

app.use(
  '/graphql',
  cors(corsOptions),
  json(),
  expressMiddleware(server, {
    context: async ({ res }) => ({ res, prisma }),
  })
)

app.listen(PORT, () => {
  console.log(`Server is running on port: ${PORT}`)
})
