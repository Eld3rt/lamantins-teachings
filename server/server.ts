import express, { json } from 'express'
import cors from 'cors'
// import gql from 'graphql-tag'
import { PrismaClient } from '@prisma/client'
import { ApolloServer } from '@apollo/server'
import { expressMiddleware } from '@apollo/server/express4'
import { resolvers } from './src/resolvers.js'
import { readFileSync } from 'fs'

const PORT = process.env.PORT || 5050
const app = express()

const corsOptions = {
  origin: process.env.BASE_URL,
  credentials: true,
}

app.use(cors(corsOptions))
app.use(express.json())

const typeDefs = readFileSync('./src/schema.graphql', 'utf8')

export interface MyContext {
  prisma: PrismaClient
}

const server = new ApolloServer<MyContext>({
  typeDefs,
  resolvers,
})

await server.start()

app.use(
  '/graphql',
  cors(corsOptions),
  json(),
  expressMiddleware(server, {
    context: async () => {
      const prisma = new PrismaClient()
      return {
        prisma,
      }
    },
  })
)

app.listen(PORT, () => {
  console.log(`Server is running on port: ${PORT}`)
})
