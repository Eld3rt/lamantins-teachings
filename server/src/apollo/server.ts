import { ApolloServer } from '@apollo/server'
import { PrismaClient } from '@prisma/client'
import { Response } from 'express'
import { ServerResponse } from 'http'
import { typeDefs, resolvers } from '../graphql/schema/index'

export interface MyContext {
  prisma: PrismaClient
  res: Response | ServerResponse
}

const server = new ApolloServer<MyContext>({
  typeDefs,
  resolvers,
})

export { server }
