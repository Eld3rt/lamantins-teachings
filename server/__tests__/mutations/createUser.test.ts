import gql from 'graphql-tag'
import { server } from '../../src/apollo/server'

type SingleGraphQLResponse<ResponseData> = {
  body: {
    kind: 'single'
    singleResult: {
      data: ResponseData
    }
  }
}

type ResponseData = {
  username: string
  email: string
}

it('should create a new user', async () => {
  let result = (await server.executeOperation<ResponseData>({
    query: gql`
      mutation {
        createUser(username: "test", email: "test@test.com") {
          user {
            id
            username
            email
          }
        }
      }
    `,
  })) as SingleGraphQLResponse<ResponseData>
  console.log(result.body.singleResult)
  expect(result.body.singleResult.data).toBeDefined()
  expect(result.body.singleResult.data).toMatchSnapshot()
})
