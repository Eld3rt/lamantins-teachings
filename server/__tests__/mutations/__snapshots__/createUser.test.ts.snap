// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create a new user 1`] = `
{
  "createUser": null,
}
`;
